File: service-1/src/main/java/com/nodelexchange/pingpong/service_1/service/PingPongService.java
package com.nodelexchange.pingpong.service_1.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.amqp.rabbit.annotation.RabbitListener;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import static com.nodelexchange.pingpong.service_1.configuration.MessageServiceConfig.PUBLISH_QUEUE;
import static com.nodelexchange.pingpong.service_1.configuration.MessageServiceConfig.SUBSCRIBE_QUEUE;

@Service
public class PingPongService {

    private static final Logger LOGGER = LoggerFactory.getLogger(PingPongService.class);

    private final RabbitTemplate rabbitTemplate;

    @Autowired
    public PingPongService(RabbitTemplate rabbitTemplate) {
        this.rabbitTemplate = rabbitTemplate;
    }

    public void serve(String message) {
        rabbitTemplate.convertAndSend(PUBLISH_QUEUE, message);
        LOGGER.info("Service 1 sent message: {}", message);
    }

    @RabbitListener(queues = SUBSCRIBE_QUEUE)
    public void receive(String message) {
        LOGGER.info("Service 1 received message: {}", message);
        serve("pong");
        try {
            Thread.sleep(10_000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        serve("ping");
    }
}

File: service-2/src/main/kotlin/com/nodalexchange/pingpong/service_2/service/PingPongService.kt
package com.nodalexchange.pingpong.service_2.service

import com.nodalexchange.pingpong.service_2.configuration.MessageServiceConfig.Companion.PUBLISH_QUEUE
import com.nodalexchange.pingpong.service_2.configuration.MessageServiceConfig.Companion.SUBSCRIBE_QUEUE
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.amqp.rabbit.annotation.RabbitListener
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.stereotype.Service

@Service
class PingPongService(private val rabbitTemplate: RabbitTemplate) {

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(PingPongService::class.java)
    }

    fun serve(message: String) {
        rabbitTemplate.convertAndSend(PUBLISH_QUEUE, message)
        LOGGER.info("Service 2 sent message: {}", message)
    }

    @RabbitListener(queues = [SUBSCRIBE_QUEUE])
    fun receive(message: String) {
        LOGGER.info("Service 2 received message: {}", message)
        serve("pong")
        try {
            Thread.sleep(10_000)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        }
        serve("ping")
    }
}
